src/promise.js

```diff
@@ -725,6 +734,10 @@ Promise.prototype._settlePromises = function () {
     var bitField = this._bitField;
     var len = BIT_FIELD_READ(LENGTH_MASK);
 
+    if (this.hadronTest !== undefined && _global.hadronTest !== this.hadronTest) {
+        throw new Error("settling promise created by: " + this.hadronTest + ", inside:" + _global.hadronTest);
```

Comment:
Just to make this a little more clear, you can add a suffix: ". Promises should resolve during the test that created them to avoid timing-related bugs."